src/utils/scrap_data.ts

```diff
@@ -0,0 +1,176 @@
+import puppeteer, { Browser, Page } from 'puppeteer';
+import { VIVINO_GRAPES_API, VIVINO_SEARCH_URL, VIVINO_EXPLORE_API, WINES_PER_PAGE, VIVINO_WINE_STYLES_API, VIVINO_FOODS_API } from '../const/const';
+import { REGIONS } from '../const/regions';
+import { WINE_TYPES } from '../const/wine_types';
+import { WineParams } from '../interfaces/params.interface';
+import { EXPLORE_API_PARAMS } from '../enums/explore_api_params.enum';
+import { Grape, Wine, WineStyle, Food } from '../interfaces/wine.interface';
+
+class ScrapData {
+  grapes: Grape[] = [];
+  wineStyles: WineStyle[] = [];
+  foods: Food[];
+
+  appendParam = (param: string, queryKey: string, list: any = []) => {
+    if (!param) {
+      return '';
+    }
+    let queryString = '';
+    const paramNames = param.split(',');
+    paramNames.forEach(name => {
+      if (list.length == 0) {
+        queryString += `&${queryKey}=${name}`;
+      } else {
+        const obj = list.find((g: any) => g.name.toUpperCase().indexOf(name.toUpperCase()) > -1);
+        if (obj) {
+          queryString += `&${queryKey}=${obj.id}`;
+        }
+      }
+    });
+    return queryString;
+  };
+
+  getVivinoSearchUrl = (pageIndex: number, wineParams: WineParams): string => {
+    wineParams.minPrice = isNaN(wineParams.minPrice) ? 0 : wineParams.minPrice;
+    wineParams.maxPrice = isNaN(wineParams.maxPrice) ? 99999999 : wineParams.maxPrice;
+    wineParams.minRating = isNaN(wineParams.minRating) ? 1 : wineParams.minRating;
+    wineParams.orderBy = !wineParams.orderBy ? 'ratings_average' : wineParams.orderBy;
+
+    let url = `${VIVINO_EXPLORE_API}?${EXPLORE_API_PARAMS.country_code}=US&${EXPLORE_API_PARAMS.currency_code}=USD&${EXPLORE_API_PARAMS.grape_filter}=varietal`;
+    url += `&${EXPLORE_API_PARAMS.order_by}=${wineParams.orderBy}&${EXPLORE_API_PARAMS.order}=desc&${EXPLORE_API_PARAMS.page}=${pageIndex}`;
+    url += this.appendParam(wineParams.grapes, EXPLORE_API_PARAMS.grape_ids, this.grapes);
```

Comment:
You are using EXPLORE_API_PARAMS and WINES_PER_PAGE.
If the site gets adjusted WINES_PER_PAGE or renamed some parameters in EXPLORE_API_PARAMS then what will happen with your algorithm?
Do you see any abilities to "alert"/"return an error" if their API will be updated? Or is it not a case for this site?

Reply:
If the site gets adjusted WINES_PER_PAGE or renamed some parameters in EXPLORE_API_PARAMS, then those params will not be considered when getting the response.

Since all the params are optional for this site, I did not find a way to return an error when the API is updated.